codegen/src/main.zig

```diff
@@ -634,7 +752,465 @@ fn generateMetadataFunction(operation_name: []const u8, state: GenerationState,
+    }
+
+    var hash_map = std.StringHashMapUnmanaged(smithy.TypeMember){};
+
+    const shape_members = getShapeMembers(shape);
+    for (shape_members) |member| {
+        try hash_map.putNoClobber(state.allocator, member.name, member);
+    }
+
+    for (shape_members) |member| {
+        for (member.traits) |trait| {
+            switch (trait) {
+                .http_header, .http_query => {
+                    std.debug.assert(hash_map.remove(member.name));
+                    break;
+                },
+                else => continue,
+            }
+        }
+    }
+
+    if (hash_map.count() == 0) {
+        return null;
+    }
+
+    var json_members = std.ArrayListUnmanaged(JsonMember){};
+
+    var iter = hash_map.iterator();
+    while (iter.next()) |kvp| {
+        const member = kvp.value_ptr.*;
+
+        const key = blk: {
+            if (findTrait(.json_name, member.traits)) |trait| {
+                break :blk trait.json_name;
+            }
+
+            break :blk member.name;
+        };
+
+        try json_members.append(allocator, .{
+            .field_name = try constantName(allocator, member.name, .snake),
+            .json_key = key,
+            .target = member.target,
+            .type_member = member,
+            .shape_info = try shapeInfoForId(member.target, state.file_state.shapes),
+        });
+    }
+
+    return json_members;
+}
+
+fn generateToJsonFunction(shape_id: []const u8, writer: std.io.AnyWriter, state: GenerationState, comptime options: GenerateTypeOptions) !void {
+    _ = options;
+    const allocator = state.allocator;
+
+    const shape_info = try shapeInfoForId(shape_id, state.file_state.shapes);
+    const shape = shape_info.shape;
+
+    if (try getJsonMembers(allocator, shape, state)) |json_members| {
+        if (json_members.items.len > 0) {
+            try writer.writeAll("pub fn jsonStringify(self: @This(), jw: anytype) !void {\n");
+            try writer.writeAll("try jw.beginObject();\n");
+            try writer.writeAll("{\n");
+
+            for (json_members.items) |member| {
+                const member_value = try getMemberValueJson(allocator, "self", member);
+                defer allocator.free(member_value);
+
+                try writer.print("try jw.objectField(\"{s}\");\n", .{member.json_key});
+                try writeMemberJson(
+                    .{
+                        .shape_id = member.target,
+                        .field_name = member.field_name,
+                        .field_value = member_value,
+                        .state = state.indent(),
+                        .member = member.type_member,
+                    },
+                    writer,
+                );
+            }
+
+            try writer.writeAll("}\n");
+            try writer.writeAll("try jw.endObject();\n");
+            try writer.writeAll("}\n\n");
+        }
+    }
+}
+
+fn getShapeTraits(shape: Shape) []smithy.Trait {
+    return switch (shape) {
```

Comment:
I don't use this often, but I think an inline switch prong would be more readable (and a little less brittle) here: https://ziglang.org/documentation/0.14.1/#Inline-Switch-Prongs